Allow configuring the theme for syntax highlighting

boxofrox opened the pull request:
These patches build on @mre's initial work.  Only the default themes included by syntect are allowed.

~~I reverted @mre's changes to `Config`, `Document`, and `CodeBlock` since we only need to set the theme name once, and the `SETUP` variable already distributes themes to the `initialize_codeblock` and `decorate_markdown` functions.~~

~~I might be able to avoid the reversion with closures, but this PR is my third attempt, and I'd rather get feedback before I spend more time on this.~~

I tested the PR via the following process on Linux:

```
cd /tmp
git clone git@github.com:boxofrox/cobalt.rs.git
cd cobalt.rs
git checkout config-theme
cargo install --force cobalt-bin --path ./ --features=syntax-highlight
cd <cobalt-blog-directory>
cat "syntax-highlight:\n  theme: base16-ocean.light" >> .cobalt.yml
cobalt build
cobalt serve
# open localhost:3000 in browser
```

Use `cobalt list-syntax-themes` to view a list of available themes.

If acceptable to @mre's needs, fixes #202.  Otherwise, let's discuss what's missing.

*Updates:*

- Switch to patch set using closures instead of global `RefCell` variable.
- Syntax theme configured in `.cobalt.yml` with
    ```yaml
    syntax-highlight:
      theme: base16-eighties.dark
    ```
- Rename subcommand used to list syntax themes from `syntax-themes` to `list-syntax-themes`.
- Fix clippy lint errors.

mre commented:
Highly appreciated! Thanks @boxofrox. 😃 

>  reverted @mre's changes to Config, Document, and CodeBlock

Whoops. Sorry for the inconvenience.

boxofrox commented:
No inconvenience.  I wouldn't know where to start on this without your initial attempt.  I tried to keep that code as I prefer local/scoped variables over global vars, but my first two attempts ran into issues with `liquid::Block` being a `Fn` type instead of a trait.

Cheers

boxofrox commented:
Awww... AppVeyor doesn't like my mutable variable behind a feature gate.  :slightly_frowning_face:
Rewrite!

epage commented on src/main.rs:
```diff
@@ -204,8 +207,13 @@ fn run() -> Result<()> {
+
+    #[cfg(all(feature="syntax-highlight", not(windows)))]
+    let app_cli = app_cli.subcommand(SubCommand::with_name("syntax-themes")
+                        .about("list available themes"));
```
Would it simplify things if we went the "null object" route and made `list_themes` return nothing on Windows?

It let's us isolate the platform checks to just a few areas

boxofrox replied:
Agreed.

epage commented on src/syntax_highlight.rs:
```diff
@@ -34,7 +36,8 @@ unsafe impl Sync for Setup {}
         syntax_set: SyntaxSet::load_defaults_newlines(),
-        theme_set: ThemeSet::load_defaults()
+        theme_set: ThemeSet::load_defaults(),
+        theme_name: RefCell::new(DEFAULT_THEME_NAME.to_owned()),
```
This feels a little dirty to me.

I'll need to look at this more to see if there is another reasonable feasible way.

boxofrox replied:
It doesn't give me a warm fuzzy feeling either.

It's just simple and works for this situation.  Should we want to add to it later to, I dunno, support varying syntax themes or anything that needs to change `theme_name` more than once, then it's worth overhauling the code to do away with this.

I might also be able to rewrite this with closures, but I fear implementation details in `syntax_highlight` will leak into the `document` module since Rust closures cannot be returned from a function, afaik.  I'll see what I come up with.

See my [other comment](https://github.com/cobalt-org/cobalt.rs/issues/202#issuecomment-313758943) regarding changing `liquid::Block` into a trait as one option.  I'd be happy to submit a PR for this.

Thanks for the review.

boxofrox replied:
I found a way with closures.  You can [view the alternate patch set](https://github.com/boxofrox/cobalt.rs/commits/config-theme-closure-example) or [compare against this PR](https://github.com/boxofrox/cobalt.rs/compare/config-theme...boxofrox:config-theme-closure-example).  With closures I was able to drop the reversion path and `RefCell`.

Let me know if the alternate is preferred and I'll swap it into this PR.

epage replied:
Heh, no wonder you haven't replied to me, I forgot to hit submit...

boxofrox replied:
:grin:

My pros (:+1:) and cons (:-1:) are pretty subjective.

The alternate patch set...

1.  :+1: Avoids using a global variable `SETUP.theme_name` in favor of scoped variable.
2.  :+1: Simplifies the PR by dropping two patches.
4.  :+1: Puts `theme` in `Config` (I'm assuming this continues to keep all config file options within one struct).
5.  :neutral_face: Wraps `initialize_codeblock()` in a closure.
6.  :neutral_face: No `RefCell`.  I'm not sure why this would matter.

The original patch set...

1.  :+1: Keeps `theme_name` organized more closely with relevant code in `syntax_highlight.rs`.
2.  :-1: Uses the global variable mentioned above.
3.  :-1: Reverts quite a bit of @mre's initial work (which was heading in the direction I favored).
4.  :-1: Stores a config file option separate from `Config` struct.
5.  :neutral_face: Works now, but global vars tend to cause problems later.

My OCD prefers the alternate patch set.  I would have submitted it instead had I figured it out sooner.

epage replied:
Sounds like we should go with the alternative then :)

Thanks for all your work on this!

boxofrox replied:
My pleasure.  Swapping out the patches now.

boxofrox commented:
Rebased all patches and applied `cargo fmt` to each using `rustfmt` v0.8.3.

boxofrox commented:
Minor edit on [18c7e97](https://github.com/cobalt-org/cobalt.rs/pull/259/commits/18c7397f5292c6d21e0762d2ba6ddb7373dc6360#diff-639fbc4ef05b315af92b4d836c31b023R411).  `cobalt syntax-themes` will use `println!` instead of `info!` for use in shell scripts, should anyone desire such.

boxofrox commented:
One point of concern.  I named the CLI subcommand `syntax-themes` thinking it may be possible to have themes for some other aspect of the blog.  But the yaml setting was left as `theme`.  If some other component of a cobalt blog becomes "theme-able", then renaming the setting to `syntax_theme`, or some such, may cause problems for users.

Maybe this can wait until we cross that bridge, maybe it doesn't matter.  Blogging is unfamiliar territory for me.

What say you?

epage commented:
> One point of concern. I named the CLI subcommand syntax-themes thinking it may be possible to have themes for some other aspect of the blog. But the yaml setting was left as theme. If some other component of a cobalt blog becomes "theme-able", then renaming the setting to syntax-theme, or some such, may cause problems for users.

> Maybe this can wait until we cross that bridge, maybe it doesn't matter. Blogging is unfamiliar territory for me.

I think it'd be good to have its name more clearly describe what it means.

I'm "almost done" with my change to using serde for documents which includes some breaking changes (#257).  Cobalt config will go through a similar clean up afterwards.  So the main value of settling on a term now is less likely for people to break (I'm tempted to write migration tools though to minimize the pain though, it'll just slow down my PR even further :( ).

uwearzt commented:
Hi,

maybe we should prefix with the highlighting engine? Something like ```syntax-syntect-theme```, there is already a issue for adding different theming engines. Then we would need a config like
```
syntax-highlighting=syntect
syntax-syntect-theme=darcula
```

boxofrox commented:
It is yaml.  How about:

```yaml
syntax-highlight:
  engine: syntect
  theme: darcula
```

uwearzt commented:
Yep, that looks nice :),

@boxofrox Sorry I was confused by https://github.com/cobalt-org/cobalt.rs/issues/200, but this one refers to highlight vim syntax, not with vim.

boxofrox commented:
@uwearzt ah, ok.  In that case, it appears we're sticking with only syntect for now.

For now, let's leave out the `engine` field.  Should cobalt later support other highlight engines, it'll be trivial to add it, I think.

```yaml
syntax-highlight:
  theme: base16-eighties.dark
```

As for custom themes, I haven't figured out how syntect handles that.  If I figure it out, I'll submit a PR to support them.  My plan there is if `theme` is a path to a file, then load it into the `SyntaxSet`, otherwise, try to use `theme` as name for default themes, and finally, fall back to 'base16-ocean.dark'.  For example:

```yaml
syntax-highlight:
  theme: syntax/darcula.tmTheme
```

epage commented:
> As for custom themes, I haven't figured out how syntect handles that. If I figure it out, I'll submit a PR to support them. My plan there is if theme is a path to a file, then load it into the SyntaxSet, otherwise, try to use theme as name for default themes, and finally, fall back to 'base16-ocean.dark'. For example:

Don't wait on perfection; it can always be added in a later PR>

As for how it should behave, I was thinking that we could have a `syntax` directory under the `_cobalt_ directory we've been talking in #211.  Under that could be a `syntax` directory.  Those syntases could act as if they are built-in to syntect.  Maybe if a theme name isn't in syntect, we append an extension and load it from this directory.

boxofrox commented:
> Don't wait on perfection; it can always be added in a later PR

Certainly.  I haven't made any changes to this patch set regarding the yaml discussion because I wasn't sure whether I was churning patches too much as it is for review purposes.  I wanted to get your thoughts, @epage, on my [previous comment](https://github.com/cobalt-org/cobalt.rs/pull/259#discussion_r126252855) before I attached the yaml patch.  Then again, I probably misunderstood what you meant by that statement.

> Maybe if a theme name isn't in syntect, we append an extension and load it from this directory.

Will do.

boxofrox commented on src/cobalt.rs:
```diff
@@ -190,7 +190,7 @@ pub fn build(config: &Config) -> Result<()> {
 
         let mut context = doc.get_render_context(&posts_data);
-        let doc_html = try!(doc.render(&mut context, source, &layouts, &mut layouts_cache));
+        let doc_html = try!(doc.render(&mut context, &source, &layouts, &mut layouts_cache));
```
This change is squashed across two commits.

Bad revert on my part.  @mre had appended a `&config` parameter and switched `source` to a reference.  `&config` wasn't necessary if using a global variable to store the theme name, and I failed to restore `&source` to `source`.

If we go with the alternate patch set, then the reversion doesn't happen.  If you prefer to keep `source` over `&source`, I'll need to update that patch set.

epage replied:
iirc clippy will complain about a useless `&`

boxofrox commented:
Alternate patch set is ready to go! (pending CI)

- Use closures instead of global RefCell variable.
- Syntax theme configured in .cobalt.yml with
    ```yaml
    syntax-highlight:
      theme: base16-eighties.dark
    ```
- Rename subcommand used to list syntax themes from `syntax-themes` to `list-syntax-themes`.
- Fix clippy lint errors.

epage commented on src/config.rs:
```diff
@@ -133,6 +138,18 @@ impl Config {
+                }
+            };
+        }
+
```
I think i'd rather isolate the feature knowledge by having this run unconditionally.

In thinking of the case where someone is developing a blog on windows and deploying with TravisCI, it'd be nice if `has_theme` unconditionally returned true.  This is a bit of a lie but then again, no validation is performed otherwise because of the feature toggle.

boxofrox replied:
TravisCI will catch the `warn!` and fail?  If not, then I'd rather `has_theme` returns false and generates the non-fatal warning.

I'm opposed to sending mixed signals to users.  If a Windows user sets the `theme` field in `.cobalt.yml`, I don't want to give them an indication that cobalt used their syntax theme and then they see the code blocks on their site aren't highlighted.

Another option is to change the `warn!` to `info!` if TravisCI complains about the former and ignores the latter.  It reduces the visibility of the message, but it's a trade-off worth making if the user is informed with cobalt's actual behavior.

If neither option is suitable, I'll defer and update `has_theme` to return false.

boxofrox replied:
Possible fix in commit 'Isolate platform checks for `has_syntax_theme`'.

There's a bit more code to it, but I like the explicitness.  If the platform is supported, return `Ok(bool)`, otherwise return an `Err(reason)`.  An additional benefit is the user is notified why an unsupported feature they're trying to use doesn't work.  Using an `Option` reduces the amount of code in the patch, but seemed like a [possibly] appropriate use of `Option`.  `Option` was definitely less explicit in it's intent.

Overall, the commit is not as succinct as returning a `vec![]` was with `list_syntax_themes`, but I wanted to put this implementation on the table in case you find it preferable.  If so, I'm open to refactoring `list_syntax_themes` to use `Result` for consistency.

epage replied:
Agreed about not giving windows users false expectations. There is a tension between conveying accurate information and allowing a linux-deployed blog be written on Windows when the supported features are different.

While I've not yet looked at the diff, I wonder if I was unclear about my suggestion. If syntax highlighting is unsupported, that is when you ignore the value. That allows someone to write their blog on windows with successful builds (I consider WSL a cop out). They then get their feedback on themes when the build fails or succeeds on Travis. Logging a message when a theme is set on a platform with no syntax highlighting seems like a good compromise if a windows developer's expectations aren't met to help them realize what is going on.

epage commented on src/main.rs:
```diff
@@ -204,8 +207,14 @@ fn run() -> Result<()> {
+    #[cfg(all(feature="syntax-highlight"))]
+    let app_cli =
+        app_cli
+            .subcommand(SubCommand::with_name("list-syntax-themes").about("list available themes"));
```
You made list_syntax_themes run regardless of whether the feature is supported or not, so why not take advantage of it and remove the feature toggles here?

boxofrox replied:
I think my reasoning at the time was without `features=syntax-highlight`, then:

1.  `list-syntax-themes` is an invalid subcommand,
2.  don't not show `list-syntax-themes` when `cobalt help` runs,
3.  bail in the subcommand match to mimic invoking the default match arm when the user invoked `list-syntax-themes` as an invalid subcommand.

Though (3) doesn't bail at the moment since *clap* generates an error for invalid subcommands.

Sounds like you don't mind `list-syntax-themes` being an available subcommand regardless whether the build includes *syntect*, so I can drop the feature toggle at [212](https://github.com/cobalt-org/cobalt.rs/pull/259#discussion-diff-126850433R212), and instead of mimicking the default subcommand behavior, bail with a message on stderr like "No themes available.  This build of cobalt does not include support for syntax highlighting".

boxofrox replied:
Fixed in commit 'Remove feature-gate on `list-syntax-themes` subcommand'.

epage replied:
I think its a matter of perspective. Is syntax highlighting optional because people want to turn it off or because it is platform specific?

I don't know the implementer's intent and so I lean towards platform specific. The benefit to it being a dedicated feature are platforms or environments where the external dep isn't as easily available and to make the conditional nature of this feature more explicit.

epage commented on src/syntax_highlight.rs:
```diff
@@ -41,33 +39,33 @@ lazy_static!{
 
 impl<'a> DecoratedParser<'a> {
-    pub fn new(parser: Parser<'a>) -> Self {
+    pub fn new(parser: Parser<'a>, config: &'a Config) -> Self {
```
I'm a little iffy on `Config` being passed everywhere.

It starts to turn it into a god object which will make unit testing more difficult.  I know it saved a little on parameter lists with `excerpt_separator` but in that case I'm actually removing that in change I'm preparing to post.

boxofrox replied:
Fixed in commit 'Remove use of Config as god object'.

boxofrox commented:
> This is looking great! Thanks for pushing a little harder on finding a way to make this work!

My pleasure.  Thanks for taking the time to review and point out areas of improvement.

I went through the patches and renamed `has_theme` to `has_syntax_theme`.  Seemed inconsistent to have `list_syntax_themes` and `has_theme`.  Also tried to wrap long strings by sentence using `concat!`.

With the last three commits of this PR, I've addressed your second and third concerns directly.  The last patch pertains to your first concern with `list_syntax_themes`.  It demonstrates a solution using `Result<bool>` to handle unsupported platforms, as I mentioned in my other reply.  If you still prefer returning false instead, I'll replace the last patch and make it so.

*clippy* is clear, and *rustfmt* has two minor complaints for lines it can't wrap.

boxofrox commented:
Those two minor complains from `rustfmt` bit me.  More aggressive wrapping with `concat!`.

epage commented:
Sorry for my delayed response. I've been away from my keyboard the last three days. There is a small chance that I'll get to look at it tonight, otherwise it'll be Monday at best (but I'll push hard to make Monday work; you've put a lot of work into this and its dragged on long enough).

boxofrox commented:
Thanks for heads up, and no worries.  I'd rather do this well than rush it.  We spend time on this as we find such time not otherwise needed by other priorities in our lives.

In the meantime, all the instructions are here for anyone to try out the feature before the PR is merged.  Don't feel you have to rush on my account.

Cheers.

epage commented:
Everything looks fantastic with the one open question of how to handle windows.  I'm fine merging this as-is and having that resolved in another PR.
